src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -566,20 +596,69 @@ internal bool TryRead(out ReadResult result)
             }
         }
 
-        private static void TrySchedule(PipeScheduler scheduler, Action action)
+        private static void TrySchedule<TState>(PipeScheduler scheduler, Action<TState> action, TState state)
         {
             if (action != null)
             {
-                scheduler.Schedule(action);
+                scheduler.Schedule(action, state);
             }
         }
 
-        private static void TrySchedule(PipeScheduler scheduler, Action<object> action, object state)
+        private static void TrySchedule(PipeScheduler scheduler, SynchronizationContext synchronizationContext, ExecutionContext executionContext, Action<object> completion, object completionState)
         {
-            if (action != null)
+            // Nothing to do
+            if (completion == null)
             {
-                scheduler.Schedule(action, state);
+                return;
             }
+
+            // Ultimately, we need to call either
+            // 1. The sync context with a delegate
+            // 2. The scheduler with a delegate
+            // That delegate and state will either be the action passed in directly
+            // or it will be that specified delegate wrapped in ExecutionContext.Run
+
+            if (synchronizationContext == null)
+            {
+                // We don't have a SynchronizationContext so execute on the specified scheduler
+                if (executionContext == null)
+                {
+                    // We can run directly, this should be the default fast path
+                    scheduler.Schedule(completion, completionState);
+                    return;
+                }
+
+                // We also have to run on the specified execution context so run the scheduler and execute the
+                // delegate on the execution context
+                scheduler.Schedule(s_scheduleWithExecutionContextCallback, new CompletionData(completion, completionState, executionContext));
+            }
+            else
+            {
+                var completionData = new CompletionData(completion, completionState, executionContext);
+                if (executionContext == null)
+                {
+                    // We need to box the struct here since there's no generic overload for state
+                    synchronizationContext.Post(s_syncContextExecuteWithoutExecutionContextCallback, completionData);
+                }
+                else
+                {
+                    // We need to execute the callback with the execution context
+                    synchronizationContext.Post(s_syncContextExecutionContextCallback, completionData);
+                }
+            }
+        }
+
+        private static void ExecuteWithoutExecutionContext(object state)
+        {
+            CompletionData completionData = (CompletionData)state;
+            completionData.Completion(completionData.CompletionState);
```

Comment:
Do you call `Completion` anywhere else? Why does it need to take argument?

Reply:
Oh, it's a delegate.

Reply:
I'm confused by the question. What do you mean?

Reply:
I though it was a method on the struct and you were calling in passing in property of the same struct. Then I saw it's a delegate.

Reply:
Ah.